### DISCUSSION TITLE
Sponsorship

### DISCUSSION THREAD

**Discussion by syrusakbary** [titzer/virgil#101] [Ideas]
Hi @titzer ,

I'm great admirer Virgil and the work you are doing with wizard-engine. I was wondering if you may be interested in setting up sponsorship for the project, so we can start spreading the love from Wasmer.

I think Virgil has really great potential for the future and it might be ready for some prime time!

_Note: Wasmer already sponsors Zig, Mold and other great software!_

**Comment by titzer** [comment]
What kind of sponsorship did you have in mind?

**Comment by syrusakbary** [comment]
I was thinking on a mid-size sponsorship ($100/month, we don't have big pockets just yet!).
Hopefully with that kind of sponsorship we could help towards first-class WASI support and publishing on the Wasmer ecosystem (we can also contribute on the latter!).

But in general, the sponsorship is way more open ended and a way to show love for the project. I think Virgil has great qualities that make it ideal for a fast-to-compile/fast-to-run GC language :)

**Comment by titzer** [comment]
When I look at where gaps are, I think the limiting resource is more engineering and development time, rather than funding atm. E.g. if a volunteer were to work on WASI support in Virgil, that would help make progress on that front.

**Comment by syrusakbary** [comment]
I'd love to spend eng resources on that but unfortunately that's a bit hard for us at the moment. If I was familiar with Virgil I'd do that myself, but I still haven't reached that point.

I suggested sponsorship mainly because I wanted to cheer on the project.
In any case, if you are not interested in sponsorship I think we can just close this ticket

**Comment by titzer** [comment]
One idea might be to sponsor a project (or a contest!) that could be carried out by a student or internet denizen. So, for example, a $1000 prize for the most (fast, complete, readable) implementation of something important.

For my part I've been thinking of putting $3k-$5k of my own money to sponsor a register allocation project (actually, contest!).

**Comment by syrusakbary** [comment]
You just gave me an idea.
We did [collaborated recently with WASM-4](https://itch.io/jam/wasm4-v2) helping them with the game jam for attracting more people to the WASM-4 and Wasmer ecosystems.
https://wasm4.org/blog/jam-2-results/

Perhaps Wasmer can do something similar, sponsoring also $2,500 on this project towards this goals:
* Get Virgil running properly running in WASI (difficulty level: 2/3)
  * Virgil WASI tests should be passing
  * Aenas.wasm should be able to compile Virgil itself
* Integrate Virgil into WAPM  (difficulty level: 1/3)
  * Publish Virgil automatically to WAPM when a new version is tagged
* Make Virgil bootstrap itself from a WASI binary (similar to what [Zig is doing](https://github.com/ziglang/zig/pull/13560), but using Wizard Engine) - no need to use the JVM (difficulty level: 3/3)

The only thing is this would be a bounty more than a contest (but we are good with this, it's just a different approach)

Other thing that we can do is that someone more experience does that as the ground work (for example, we handle the wapm integration and someone else handle the wasi integration), and then we create a contest to do cool things on top of Virgil

**Comment by titzer** [comment]
That sounds like a great idea!

In general, bootstrapping on WASI should be a very doable project. The stretch goal would be to implement a lot more on top of WASI to Virgil a really good development language. I think we should do this!

**Comment by syrusakbary** [comment]
Awesome. What should be the next steps?
Happy to hop in a call if that may be helpful

**Comment by titzer** [comment]
Sure, we can talk later this week. Feel free to send me an email to set up a time.

**Comment by syrusakbary** [comment]
Tried searching around, but couldn't find your email (mine is syrus@wasmer.io)

We can try to see if any time available here would work for you: https://calendly.com/syrus-wasmer (Wednesday 21 -tomorrow- would work great for me)
